crates/core/component/dex/src/component/position_manager.rs

```diff
@@ -311,30 +361,31 @@ pub(crate) trait Inner: StateWrite {
     ///
     /// This should be the SOLE ENTRYPOINT for writing positions to the state.
     /// All other position changes exposed by the `PositionManager` should run through here.
-    #[tracing::instrument(level = "debug", skip(self, position), fields(id = ?position.id()))]
-    async fn update_position(&mut self, position: position::Position) -> Result<()> {
-        let id = position.id();
-        tracing::debug!(?position, "fetch position's previous state from storage");
-        // We pull the position from the state unconditionally, since we will
-        // always need to update the position's liquidity index.
-        let prev = self
-            .position_by_id(&id)
-            .await
-            .expect("fetching position should not fail");
+    #[tracing::instrument(level = "debug", skip_all,  fields(id = ?new_state.id()))]
+    async fn update_position(
+        &mut self,
+        prev_state: Option<Position>,
+        new_state: Position,
+    ) -> Result<()> {
+        tracing::debug!(?prev_state, ?new_state, "updating position state");
 
         // Clear any existing indexes of the position, since changes to the
         // reserves or the position state might have invalidated them.
-        self.deindex_position_by_price(&position);
+        if let Some(prev_state) = prev_state.as_ref() {
+            self.deindex_position_by_price(&prev_state);
+        }
 
         // Only index the position's liquidity if it is active.
-        if position.state == position::State::Opened {
-            self.index_position_by_price(&position);
+        if new_state.state == position::State::Opened {
+            self.index_position_by_price(&new_state);
         }
 
         // Update the available liquidity for this position's trading pair.
-        self.update_available_liquidity(&position, &prev).await?;
+        self.update_available_liquidity(&new_state, &prev_state)
```

Comment:
This PR actually improves on the previous impl because prev_state is `Option<Position>` but noting that the parent method signature is:
```rust
async fn update_position(
         &mut self,
         prev_state: Option<Position>,
         new_state: Position,
    ) -> Result<()>;
```
And this method is:

```rust
async fn update_available_liquidity(
        &mut self,
        position: &Position,
        prev_position: &Option<Position>,
    ) -> Result<()>;
```

but since the types are different this is probably fine

Reply:
I'd rather streamline that code in a later PR that implements eviction, but good spotting. Added a TODO.